src/index.ts

```diff
@@ -854,6 +863,88 @@ export abstract class PinejsClientCore<PinejsClient> {
 		}
 	}
 
+	private waitAsync(delayMs: number): Promise<void> {
+		return new Promise<void>((resolve) => {
+			setTimeout(resolve, delayMs);
+		});
+	}
+
+	protected async callWithDelay<T>(
+		fnCall: (attempts: number, isLastAttempt: boolean) => Promise<T>,
+		delayMs: number,
+		attempt: number,
+		final: boolean,
+		logAttempt: boolean,
+	): Promise<T> {
+		if (!delayMs) {
+			return await fnCall(attempt, final);
+		}
+
+		if (logAttempt) {
+			const msg = `Retrying balena api request in ${
+				delayMs / 1000
+			} seconds (attempt ${attempt})`;
+			console.log(msg);
+		}
```

Comment:
I would not directly log anything here, instead I'd make it be an optional callback so the caller can log or do whatever else they want with the info, eg
```suggestion
		logAttempt?.(delayMs, attempt)